src/hidra/receiver/asapo_transfer.py

```diff
@@ -0,0 +1,208 @@
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import argparse
+import logging
+import random
+import signal
+import socket
+from threading import Event
+from time import sleep
+import yaml
+import zmq
+
+from hidra import Transfer, __version__, generate_filepath, _constants
+from plugins.asapo_producer import AsapoWorker
+
+logger = logging.getLogger(__name__)
+
+
+class Stopped(Exception):
+    """Raised when a run is stopped."""
+    pass
+
+
+class TransferConfig:
+    def __init__(self, signal_host, target_host, detector_id,
+                 endpoint, beamline, default_data_source,
+                 token_path='/gpfs/asapo/shared/beamline_tokens',
+                 target_dir='', n_threads=1, start_file_idx=1, beamtime='auto',
+                 file_regex="current/raw/(?P<scan_id>.*)_(?P<file_idx_in_scan>.*).h5",
+                 timeout=30, reconnect_timeout=3, log_level="INFO"):
+        
+        self.detector_id = detector_id
+        self.log_level = log_level
+        
+        self.endpoint = endpoint
+        self.beamtime = beamtime
+        self.beamline = beamline
+        self.token = read_token(token_path, beamline)
+        self.n_threads = n_threads
+        self.timeout = timeout
+        self.default_data_source = default_data_source
+        self.file_regex = file_regex
+        self.start_file_idx = start_file_idx
+        
+        self.signal_host = signal_host
+        self.target_host = target_host
+        self.target_dir = target_dir
+        self.reconnect_timeout = reconnect_timeout
+
+    def __repr__(self):
+        return str(self.__dict__)
+
+
+def read_token(token_path, beamline):
+    with open(f"{token_path}/{beamline}.token", "r") as f:
+        token = f.readline().split("\n")[0]
+    return token
+
+
+def create_query(signal_host, detector_id):
+    logger.info(
+        "Creating Transfer instance type=%s signal_host=%s detector_id=%s use_log=True",
+        "STREAM_METADATA", signal_host, detector_id)
+    query = Transfer("STREAM_METADATA", signal_host, detector_id=detector_id,
+                     use_log=True)
+    return query
+
+
+def create_asapo_transfer(asapo_worker, query, target_host, target_dir, reconnect_timeout):
+    target_port = random.randrange(50101, 50200)
+    asapo_transfer = AsapoTransfer(
+        asapo_worker=asapo_worker,
+        query=query,
+        target_host=target_host,
+        target_port=target_port,
+        target_dir=target_dir,
+        reconnect_timeout=reconnect_timeout)
+
+    signal.signal(signal.SIGINT, lambda s, f: asapo_transfer.stop())
+    signal.signal(signal.SIGTERM, lambda s, f: asapo_transfer.stop())
```

Comment:
I would move these out of this function and do it in `run_transfer`.

Reply:
This is in `run_transfer` now